var/spack/repos/builtin/packages/mesa/package.py

```diff
@@ -52,10 +56,27 @@ class Mesa(AutotoolsPackage):
     depends_on('libxfixes')
 
     depends_on('glproto@1.4.14:', type='build')
-    depends_on('dri2proto@2.6:', type='build')
-    depends_on('dri3proto@1.0:', type='build')
+    depends_on('libdrm', when='+dri')
+    depends_on('dri2proto@2.6:', type='build', when='+dri')
+    depends_on('dri3proto@1.0:', type='build', when='+dri')
+    depends_on('expat', when='+dri')
+    depends_on('llvm@3:3.99', when='+dri')
```

Comment:
I did not tested +dri, with +gallium it is needed, likely for shaders, as @davydden suggest. But if I' m not wrong, dri should wrap OpenGL, so shaders parsing should be handled by the native OpenGL driver. 
Hard to tell without test on the correct environment.